src/io/aviso/taxi_toolkit/utils.clj

```diff
@@ -47,17 +48,22 @@
             (Thread/sleep 100)
             (retry f timestamp)))))))
 
-(defn retry-times
-  "Keeps retrying an action specified number of times (defaults to 10), with 100 ms delay between retries.
-Useful when an element is not available for the momen   (i.e. it's being animated or re-drawn by AngularJS scope change) and when the function
-execution takes time to perform."
-  ([f] (retry-times f 10))
-  ([f times]
-   (try
-     (f)
-     (catch Exception e
-       (if (<= times 1)
-         (throw e)
-         (do
-           (Thread/sleep 100)
-           (retry-times f (dec times))))))))
+(defn retry-till-timeout
+  "Keeps retrying an action till the specified timeout described by value in milliseconds,
+  with optionally overriden start time. Additionaly, a predicate can be passed which causes
+  a retry if action is successful, but predicate is not."
+  [timeout f & {:keys [pred start]
+                :or {pred (constantly true)
+                     start (System/currentTimeMillis)}}]
+  (let [wrapped #(try [::success (f) (pred)] (catch Exception e [::error e false]))]
+    (loop [[tag ret check] (wrapped)]
+      (cond
+        (and (= tag ::success) check)
+        ret
+        (> (System/currentTimeMillis) (+ start timeout))
+        (if (instance? Throwable ret)
+          (throw ret)
+          (throw (TimeoutException.)))
+        :else (do
+                (Thread/sleep 17)
```

Comment:
The sleep time is that small because if we've had an error here, it's quite possibly because an element had not yet been rendered/will be rerendered and the operation most likely will be successful fairly soon (approximating one frame). If the element is not yet there due to page being loaded, we're "sleeping" anyway because webdriver implicit wait so additional miliseconds won't give us anything.